alphagov/paas-cf
# [#173150927] AWS redis update manager tool

[Story](https://www.pivotaltracker.com/story/show/173150927)

What
----

Semi-automate the listing, previewing, and emailing of elasticache service updates.

Adds a CLI tool in `tools` called `aws_redis_update_manager` that uses:
- AWS API
- PaaS Accounts API
- CF API

to
- `print` (print which org managers are going to be emailed about which instances)
- `preview` (preview the format of the email which you are going to send)
- `email` (email all of the relevant people)

How to review
-------------

CI

Code review commit by commit

Dry run in your development environment:
- Sign in to CF dev env
- Create a redis instance
- Delete the line `.select { |update| update.update_action_status == 'not-applied' }` from `tools/aws_redis_update_manager/lib/elasticache_update_finder.rb` (so you don't have to wait for an update
- `./tools/aws_redis_update_manager/aws_redis_update_manager` with the relevant CLI flags

See [team manual PR](https://github.com/alphagov/paas-team-manual/pull/351) for guidance

Gotchas
-------

- FNV hash function is implemented in ruby as a util, because the gems I tried didn't work without monkey patching
- Not very ergonomic CLI

Example usage
-------

```
AWS_REGION=eu-west-1 gds aws paas-prod -- \
  ./tools/aws_redis_update_manager/aws_redis_update_manager \
  --action preview \
  --notify-api-key "$( command to get this from credhub )" \
  --paas-accounts-url https://accounts.cloud.service.gov.uk \
  --paas-accounts-username "$( command to get this from credhub )" \
  --paas-accounts-password "$( command to get this from credhub )" \
  --maintenance-window-date 'Tuesday 16 June 2020' \
  --maintenance-window-time-range '0600-0800' \
  --alt-maintenance-window-date 'Thursday 18 June' \
  --alt-maintenance-window-time-range '1800-2000' \
  --region Ireland
```

Who can review
--------------

Not @tlwr
